dspace/config/spring/api/bitstore.xml

```diff
@@ -7,7 +7,7 @@
         <property name="incoming" value="0"/>
         <property name="stores">
             <map>
-                <entry key="0" value-ref="localStore"/>
+                <entry key="0" value-ref="${assetstore.storename.primary}"/>
```

Comment:
Ok, so Do you suggest to add a new property inside the `dspace.cfg`?
Which should be its naming? @benbosman @tdonohue

Reply:
I would rather add `${assetstore.storename.primary}` to `dspace.cfg`, set it to 0 by default, and add some comments that it should be set to 1 if they want to switch to S3

That won't break the repo if there are bitstreams already

Reply:
@benbosman : I'm not sure I understand your recommendation here.  I don't think it's easy to swap between `0` and `1` stores like you are suggesting.

Currently the `1` store is commented out by default: https://github.com/DSpace/DSpace/blob/b9eb369275cffbeb9d51d40fe6104450c7df2a36/dspace/config/spring/api/bitstore.xml#L11

Unless I'm mistaken, I don't think we can enable (uncomment) that `1` store by default without causing errors....as soon as you uncomment that `<entry>` tag, I think the `s3Store` will attempt to load & that'll fail unless you actually have AWS keys configured.

I think the overall goal of this PR was to simply allow the value of the main `0` store to be swapped easily from `localStore` to `s3Store` based on whether you wanted to primarily use local storage or S3.  If you wanted to use *both* then you are correct you'd still have to modify the `bitstore.xml` file manually to add a separate `1` store with the settings you desire.

Reply:
@benbosman and @vins01-4science :  Although, I *do* now see Ben's point that this scenario is problematic..

1. If you add Bitstreams to DSpace first, they will be stored in `localStore` by default.
2. If you then change your `assetstore.storename.primary = s3Store`, then all existing Bitstreams in `localStore` will no longer be found.  This is because the primary store is switched to use S3...and all Bitstreams will be assumed to be in S3.

So, there is a bug here because the if you've already added some content, then that existing content could be lost unless you change `bistore.xml` to have *two stores* instead of one.

Reply:
Hello @benbosman and @tdonohue , thank  you for the comments.
So what I should do here?

Reply:
@vins01-4science : I'm a little stuck myself on the best direction forward.  I see @benbosman 's point now...and his solution would work _if we can make sure these beans are lazy loaded_ (or maybe`@ConditionalOnProperty`).

Ben's correct that part of the issue here is that we are NOT changing this tag `<property name="incoming" value="0"/>`.

However, we also shouldn't load the `S3BitStoreService` bean by default, as many sites don't need that bean (and it'll throw a ton warnings if you don't have the `aws*` properties specified).

My only idea is to look at what @benbosman suggests, but also make sure the S3 bean isn't loading by default.
1. First, ensure there's a config which lets you change the `incoming` property from `0` (localstore) to `1` (s3store)
2. Likely this means uncommenting this tag by default: `<entry key="1" value-ref="s3Store"/>`
3. However, make sure the `s3Store` bean does NOT load by default. this might require some testing to figure out the best approach.  Either finding a way to lazy load that bean, or maybe add a `@ConditionalOnProperty` annotation which uses a new configuration `assetstore.s3.enabled = true/false`

Overall, though, I see @benbosman 's point that this current implementation could cause major issues if you switch to S3 _after adding content into DSpace_.  This current implementation only works well in the scenario where you have a fresh/brand new DSpace installation which needs to be configured to only use S3.

Reply:
@tdonohue: thank you for the clear review, Do we still need some feedback from @benbosman or I can proceed with the changes you requested earlier:

> 1. First, ensure there's a config which lets you change the incoming property from 0 (localstore) to 1 (s3store)
>2. Likely this means uncommenting this tag by default: <entry key="1" value-ref="s3Store"/>
> 3. However, we need make sure the s3Store bean does NOT load by default. this might require some testing to figure out the best approach. Either finding a way to lazy load that bean, or maybe add a @ConditionalOnProperty annotation which uses a new configuration assetstore.s3.enabled = true/false

Am I missing something else?

Reply:
@vins01-4science : I see that @benbosman already gave a "thumbs up" response to my comment just above where I suggested those 3 changes.  So, I think you can proceed with making those three changes to help move this PR forward.  

One minor additional suggestion, is that I think we may want to rename the config, as `assetstore.storename.primary` implies it should be a *name* of a store.  Now we are talking about switching this to the store *number*.  So, we may want to rename it to be something like `assetstore.index.primary` and documenting that setting it to `0` means it will use local storage, while setting it to `1` means S3 (and note that additional indexes may be added to bitstore.xml as needed).

Thanks!

Reply:
@tdonohue : Resolved and introduced a secondary store as backup while trying temp config with another store as the primary one. 
Also introduced new status for stores `disabled` and `initialized` just to lazy initialize them (if disabled) at the first use.
Let me know if there is something else.
Thank you!